Fixes #45983: Add gluster_heal_facts module for Gluster storage.

devyanikota opened the pull request:
This module will add support to extract self-heal status of a cluster
and return the number of files that are still in progress for the
healing process to complete.

Signed-off-by: Devyani Kota <dkota@redhat.com>

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Fixes #45983 : Adds gluster_heal_facts module
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
storage/glusterfs/gluster_heal_facts

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```
ansible 2.4.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/devyani/work/gdeploy/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.13 (default, Dec  1 2017, 09:21:53) [GCC 6.4.1 20170727 (Red Hat 6.4.1-1)]

```

##### ADDITIONAL INFORMATION
In a cluster, when a brick goes offline and comes back online, self-heal needs to be
triggered to bring all the replicas in sync.
This module will help list the files that still need to be healed, and the ones that have healed.
gluster volume heal test-volume info
```
Brick server1:/glusterfs/test-volume_0
Number of entries: 0
 
Brick server2:/glusterfs/test-volume_1
Number of entries: 101 
/95.txt
/32.txt
/66.txt
/35.txt
```

Commit:
Fixes #45983: Add gluster_heal_facts module for Gluster storage.

This module will add support to extract self-heal status of a cluster
and return the number of files that are still in progress for the
healing process to complete.

Signed-off-by: Devyani Kota <dkota@redhat.com>

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,130 @@
+DOCUMENTATION = '''
+---
+module: gluster_heal_facts
+short_description: Retrieve facts about one or more Gluster hosts
```
This should be: "Gather facts about self-heal status"

It is not facts about Gluster hosts.

devyanikota replied:
Done.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,130 @@
+author: "Devyani Kota (@devyanikota)"
+version_added: "2.8"
+description:
+  - Retrieve facts about one or more Gluster hosts.
```
Same as above

devyanikota replied:
Done.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,130 @@
+      - The volume name.
+    required: true
+    aliases: ['volume']
+  heal_type:
```
heal_type key is misleading. Can you change it to: `filter' (The filter can be of self-heal or rebalance)

devyanikota replied:
@sac 'filter' happens to be a keyword, changing it to 'status_filter'. Hope that looks ok?

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,130 @@
+    default: "self-heal"
+    choices: ["self-heal", "rebalance"]
+    description:
+      - Determines which status is to be returned. If the heal_status is
```
s/status is/facts are/

Since we are returning the facts, which also includes status.

devyanikota replied:
Done.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,130 @@
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+
+glusterbin = ''
```
Where is this value set?

devyanikota replied:
Rectified. Thanks.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,130 @@
+
+
+def get_rebalance_status(name):
+    out = run_gluster(['volume', 'rebalance', name, 'status'])
```
The data has to be parsed, and the facts have to be sorted out.
Like:

* Number of files pending heal.
* Number of files to be rebalanced
etc based on the data thrown by the heal status.

Just dumping the string as is, does not add any value to the facts module.
The end user should be able to see particular keys and make any decision. 
This way, we are shifting the responsibility of parsing the data on the caller.

sac replied:
@devyanikota you have not addressed this comment.

devyanikota replied:
@sac yes, working on it. Seems like the dictionary of bricks is not enough to allow the caller to re-use the output. Trying to extract the data from the tabular format to a mapped dictionary to allow reusability.

devyanikota replied:
@sac modified the output, returns a dictionary of the host and the number of files that need healing.
Please review the same. Thanks.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,130 @@
+        still in process are returned. If heal_status is rebalance,
+        rebalance status is returned.
+requirements:
+  - GlusterFS > 3.2
```
Where are we checking if the GlusterFS version is indeed > 3.2
Refer the peer module.

devyanikota replied:
Rectified. Added checks for GlusterFS versions.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,166 @@
+    # If the installed version is less than 3.2, it is an invalid version
+    # return True
+    return LooseVersion(version) < LooseVersion(required_version)
+
```
Can you move this function above main() as per ansible guildelines:

" 
Ansible follows C-style code flow where the caller functions/methods are towards the bottom of the file and the callee implementations are above them.
"

devyanikota replied:
Done.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,166 @@
+                         )
+    module = AnsibleModule(argument_spec)
+
+    global glusterbin
```
Can you move the global declarations to one place? Easy for the future maintainer.

devyanikota replied:
Done.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,166 @@
+def main():
+    global module
+    argument_spec = dict(name=dict(type='str', required=True, aliases=['volume']),
+                         status_filter=dict(default='self-heal', choices=['self-heal', 'rebalance']),
```
Can you make the parameters one per line?

See: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/storage/glusterfs/gluster_volume.py#L447

devyanikota replied:
Done.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,166 @@
+
+
+def get_self_heal_status(name):
+    out = run_gluster(['volume', 'heal', name, 'info'])
```
Since you're screenscraping, set the locale to C before scraping.

See: gluster_peer

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,166 @@
+        rc, out, err = module.run_command(args, **kwargs)
+        if rc != 0:
+            module.fail_json(msg='error running gluster (%s) command (rc=%d): %s' %
+                             (' '.join(args), rc, out or err), exception=traceback.format_exc())
```
If rc != 0 it is always err. You don't need out or err.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,166 @@
+
+def get_rebalance_status(name):
+    out = run_gluster(['volume', 'rebalance', name, 'status'])
+    return out
```
You have not addressed rebalance?

Commit:
Updates #45983: Add support for rebalance_status for Gluster storage.

Added support to extract rebalance status of a cluster and return the
number of files that are still in progress for the rebalance process
to complete.

Signed-off-by: Devyani Kota <dkota@redhat.com>

Akasurde commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2016 Red Hat, Inc.
```
```suggestion
# Copyright: (c) 2016, Red Hat, Inc.
```

Akasurde commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,183 @@
+    default: "self-heal"
+    choices: ["self-heal", "rebalance"]
+    description:
+      - Determines which facts are to be returned. If the status_filter is
```
```suggestion
      - Determines which facts are to be returned.
      - If the C(status_filter) is C(self-heal), status of self-heal, along with the number of files still in process are returned.
      - If the C(status_filter) is C(rebalance), rebalance status is returned.
```

Akasurde commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,183 @@
+'''
+
+EXAMPLES = '''
+# Gather facts about all gluster hosts in the cluster and return self-heal
```
```suggestion
- name: Gather self-heal facts about all gluster hosts in the cluster
  gluster_heal_facts:
    name: test_volume
    status_filter: self-heal
  register: self_heal_status
- debug:
    var: self_heal_status
```

Akasurde commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,183 @@
+
+
+def get_self_heal_status(name):
+    out = run_gluster(['volume', 'heal', name, 'info'])
```
```suggestion
    out = run_gluster(['volume', 'heal', name, 'info'], environ_update=dict(LANG='C', LC_ALL='C', LC_MESSAGES='C'))
```

Akasurde committed:
Apply suggestions from code review

Co-Authored-By: devyanikota <divs.passion.18@gmail.com>

devyanikota commented:
@sac @Akasurde hello. I have addressed the comments. Could you guys please re-review.
Thank you.

Akasurde commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,191 @@
+      - Determines which facts are to be returned.
+      - If the C(status_filter) is C(self-heal), status of self-heal, along with the number of files still in process are returned.
+      - If the C(status_filter) is C(rebalance), rebalance status is returned.
+        self-heal, status of self-heal, along with the number of files still
```
This description seems redundant to me.

devyanikota replied:
@Akasurde removed the redundant description. Please review.

Akasurde reviewed (changes requested):
Rest LGTM

sac-urs commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,191 @@
+        elif status_filter == "rebalance":
+            rebalance_status = get_rebalance_status(volume_name)
+    except Exception:
+        module.fail_json(msg="Invalid heal status option.")
```
In the case of exception print the exception e. Getting self-heal status can fail for various reasons, this error message is wrong.
If the status option is wrong ansible takes care of printing it, since we have provided possible options.

devyanikota replied:
@sac-urs I have added support to handle other possible glusterd cases, and display relevant exception errors.

sac-urs commented on lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py:
```diff
@@ -0,0 +1,191 @@
+            br_dict['status'] = line.split(":")[1].strip()
+        elif 'Number' in line:
+            br_dict['no_of_entries'] = line.split(":")[1].strip()
+            if int(br_dict['no_of_entries']) != 0:
```
This will throw an exception every time when a server is down. When glusterfsd is down the output looks like this:

Brick 10.70.43.200:/mnt/engine
Status: Transport endpoint is not connected
Number of entries: -

And you'll be trying to do int('-') which will throw ValueError. And the module throws error:
fatal: [10.70.42.25]: FAILED! => {"changed": false, "msg": "Invalid heal status option."}                                       
in the function main.

devyanikota replied:
done.

devyanikota committed:
Updates #45983: Display relevant error incase of Glusterd anomalies

Signed-off-by: Devyani Kota <devyanikota@gmail.com>

devyanikota commented:
@sac @sac-urs @Akasurde updated the PR with the suggested changes. Can you guys please re-review. Thank you.

Akasurde reviewed (approved):
LGTM

devyanikota committed:
Updates #45983: Add relevant examples for rebalance operation.

Signed-off-by: Devyani Kota <devyanikota@gmail.com>

devyanikota committed:
Updates #45983: Allow excluding the file names from heal status.

Signed-off-by: Devyani Kota <devyanikota@gmail.com>

devyanikota committed:
Updates #45983: Display only relevant information for rebalance status.

Signed-off-by: Devyani Kota <devyanikota@gmail.com>

devyanikota committed:
Updates #45983: Remove hanging dicts to allow reusability of heal status.

Signed-off-by: Devyani Kota <devyanikota@gmail.com>

sac reviewed (approved):
LGTM

sac commented:
shipit

Akasurde commented:
@devyanikota Thanks for the contribution. @sac Thanks for the reviews.
